Adds Functionality for controlling ui with serial and hardware inputs

cosparks opened the pull request:
### New Classes
#### HardwareInput
- abstract class which provides a means of linking an input (button, encoder, serial, etc..) with a callback function
- classes which implement HardwareInput will check their respective i/o every control loop and call their input callback function when input is detected
- ```input_data_t``` (which is ```int32_t```) represents the input / change in input and is passed as argument to the input callback

#### InputManager
- provides interface for assigning input callbacks to inputs
- because input behavior will change between different states/menus, we will frequently need to reassign callbacks associated with each input
- Inputs are also identified by unique id in settings.h
 
#### ControllerBase
- base class for all ui controller types
- provides implementations of all ui input handler methods, which do nothing by default (this means that ui controllers will only have to implement input handlers for the inputs which that use -- methods which aren't implemented will safely do nothing)
- expose inFocus getter, which returns index of ui element which is currently focussed

#### ControllerMenu
- default menu controller class which will define most of the functionality we need for most simple UI menus

#### ControllerFactory
- deals with nasty circular dependencies between ApplicationContext and children of ControllerBase class
- handles construction of different ui Controllers, also currently contains button configurations for different MenuController

#### Application Context
- class which handles changes in context between different menus
- named after 'context' from [state pattern](https://en.wikipedia.org/wiki/State_pattern)

### Extras
- InputSerial class adds means of controlling menu focus/selection with up/down arrow keys and return key: open serial monitor and press keys to experiment
- Converts UIEventHandler to singleton class

cosparks committed:
Adds InputManager, HardwareInput and MenuControllerExample
- still in process of fleshing out controller system
- InputManager provides ability to assign arbitrary input to arbitrary callback
- HardwareInput provides interface by which InputManager can control and assign hardware inputs to their respective callbacks
- MenuControllerExample attempts to isolate control of ui from display -- wip

cosparks committed:
Resolves merge conflicts with dev-style

cosparks committed:
Completes preliminary design for application control
- adds ApplicationContext, which handles changes between different states
- adds MenuControllerExample, which is an example of a class which defines a program state
- adds methods to MenuViewExample which make it controllable from the above mentioned class

cosparks committed:
Adds input_data_t typedef, ApplicationContext::StateData struct
- input_data_t typedef for hardware inputs
- StateData for ApplicationContext state transitions

cosparks committed:
Undoes minor changes to comments in Handleable.h

cosparks committed:
local changes ahead of remote

cosparks committed:
Fixes conflict with remote

cosparks committed:
Updates settings to include pcb pins for LCD

cosparks commented:
closes #34

LaPommeCosmique commented on src/System/HardwareInput.h:
```diff
@@ -0,0 +1,78 @@
+        InputCallback _action = [](int32_t) { }; // default action
+};
+
+class HardwareDemuxer {
```
would you be able to add a comment explaining what this class does? Its not immediately clear to me

cosparks replied:
For sure.  Have you worked with multiplexers before?  They allow you to split up signals from one to many.  We were running low on pins for all the different functions of dyno, so i added multiplexer for all the ui button inputs which allows us to read 4 buttons while only using 3 pins

cosparks replied:
I actually wound up removing that class from this PR.  I've got another PR ready to go for when this one is done which concerns reading the demuxxed inputs

LaPommeCosmique replied:
Ahh okay. So the multiplexer would be a hardware component that is represented by this HardwareDemuxer class?

cosparks replied:
exactly

cosparks committed:
Adds documentation and cleans up code
- adds comments for HardwareInput and HardwareInputSerial
- removes HardwareDemux class, which shouldn't be part of current pr

cosparks committed:
Adds ControllerBase class
- provides default implementations of all virtual input handler methods

cosparks committed:
Revision to ApplicationContext and Controller classes (details in PR)

cosparks committed:
Adds code to handle dual-core race conditions during state transitions
- ApplicationContext now performs state transitions in handle()
- StateData struct _nextState is set to the desired transition state
- When all state-transition-related rendering actions are complete, second core sets
flag in ApplicationContext to indicate that it is safe to delete old resources
- UIEventHandler now only handles one event per control loop

cosparks commented:
@LaPommeCosmique the scope of this PR has definitely ballooned beyond what I originally intended.  I've updated the notes to include the more recent changes I've made.

Sorry about that

cosparks commented:
I'm going to have another look tonight to make sure I didn't do anything stupid this morning, then I'll merge and close

cosparks committed:
Improves name for changeState bool and related methods
